File: c/tests/test_obd_dtc_row_update_0x64258.py
#!/usr/bin/env python3
"""
Verify obd_dtc_row_update_0x64258 (0x64258) against the ACTUAL ROM bytes,
run in the SH-2E emulator.

Side-effect leaf over the OBD DTC table (base 0xFFFF8930, stride 0x34,
row index = word@0xFFFF8D74):

  row = word@0xFFFF8D74
  p   = 0xFFFF8930 + (row & 0xFFFF) * 0x34
  byte@p+0x32 = (byte@p+0x32 + byte@p+0x07 + 0xFF) & 0xFF
  byte@p+0x07 = 1
  byte@p+0x32 = (byte@p+0x32 + byte@p+0x08 + 0xF9) & 0xFF
  byte@p+0x08 = 7

C:
  void obd_dtc_row_update_0x64258(void)

Run from repo root:  python3 c/tests/test_obd_dtc_row_update_0x64258.py [N]
"""
import os, sys, random

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(ROOT, 'tools'))
from sh2emu import SH2

ROM = os.path.join(ROOT, 'roms', 'stock', '60E1D400.bin')
ENTRY = 0x0064258

ROW_ADDR = 0xFFFF8D74   # 16-bit row index (mov.w read)
BASE = 0xFFFF8930       # DTC table base
STRIDE = 0x34


def model(row, b32, b07, b08):
    p = (BASE + (row & 0xFFFF) * STRIDE) & 0xFFFFFFFF
    b32 = (b32 + b07 + 0xFF) & 0xFF
    b07 = 1
    b32 = (b32 + b08 + 0xF9) & 0xFF
    b08 = 7
    return p, b32, b07, b08


def addr(row, off):
    """Effective byte address, 32-bit wrapped like the SH-2 (p + off) & 0xFFFFFFFF."""
    p = (BASE + (row & 0xFFFF) * STRIDE) & 0xFFFFFFFF
    return (p + off) & 0xFFFFFFFF


def main():
    N = int(sys.argv[1]) if len(sys.argv) > 1 else 20000
    rom = open(ROM, 'rb').read()
    cpu = SH2(rom)

    def run_one(row, b32, b07, b08):
        ram = {ROW_ADDR: (row >> 8) & 0xFF, ROW_ADDR + 1: row & 0xFF}
        for off, v in ((0x32, b32), (0x07, b07), (0x08, b08)):
            ram[addr(row, off)] = v
        cpu.call(ENTRY, ram=ram)
        return (cpu.ram.get(addr(row, 0x32), -1),
                cpu.ram.get(addr(row, 0x07), -1),
                cpu.ram.get(addr(row, 0x08), -1))

    # Targeted: corner bytes + rows 0..3.
    vals = [0x00, 0x01, 0x06, 0x07, 0x08, 0x7F, 0x80, 0xFF]
    for row in range(4):
        for b32 in vals:
            for b07 in vals:
                for b08 in vals:
                    got = run_one(row, b32, b07, b08)
                    exp = model(row, b32, b07, b08)[1:]
                    if got != exp:
                        print("FAIL: row=%d (b32,b07,b08)=(%02X,%02X,%02X) -> %s expected %s" % (
                            row, b32, b07, b08, got, exp))
                        sys.exit(1)

    # Random: realistic row indices 0..0x14 (the DTC table is 21 rows —
    # 0xFFFF8930 + 21*0x34 == 0xFFFF8D74, the row-index word itself; larger
    # rows wrap p into ROM code, which is not a real configuration).
    for _ in range(N):
        row = random.randint(0, 0x14)
        b32, b07, b08 = (random.randint(0, 255) for _ in range(3))
        got = run_one(row, b32, b07, b08)
        exp = model(row, b32, b07, b08)[1:]
        if got != exp:
            print("FAIL: row=%d (b32,b07,b08)=(%02X,%02X,%02X) -> %s expected %s" % (
                row, b32, b07, b08, got, exp))
            sys.exit(1)

    print("OK  obd_dtc_row_update_0x64258 @0x%04X  (targeted + %d random)" % (ENTRY, N))
    sys.exit(0)


if __name__ == '__main__':
    main()

File: c/tests/test_obd_dtc_row_update_0x64258.c
/* test_obd_dtc_row_update_0x64258.c
 *
 * Host C companion for obd_dtc_row_update_0x64258 (0x64258).
 *
 * The lift updates the active DTC-table row: row index word at 0xFFFF8D74,
 * table base 0xFFFF8930, stride 0x34.  Both SFRs live in page 0xFFFF8000,
 * mapped MAP_FIXED.  Rows tested are 0..3 so p stays inside the page.
 *
 * Reference: byte32 = (byte32 + byte07 + 0xFF) & 0xFF; byte07 = 1;
 *            byte32 = (byte32 + byte08 + 0xF9) & 0xFF; byte08 = 7.
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <unistd.h>

extern void obd_dtc_row_update_0x64258(void);

#define ROW_ADDR 0xFFFF8D74u
#define BASE     0xFFFF8930u
#define STRIDE   0x34u

static void map_page(uintptr_t addr)
{
    long page = sysconf(_SC_PAGESIZE);
    uintptr_t base = addr & ~((uintptr_t)page - 1);
    void *p = mmap((void *)base, (size_t)page, PROT_READ | PROT_WRITE,
                   MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1, 0);
    if (p == MAP_FAILED) {
        perror("mmap");
        exit(1);
    }
}

static void set_row(uint16_t row)
{
    *(volatile uint16_t *)ROW_ADDR = row;
}

static int check(uint16_t row, uint8_t b32, uint8_t b07, uint8_t b08)
{
    uint8_t *p = (uint8_t *)(BASE + (uint32_t)row * STRIDE);
    set_row(row);
    p[0x32] = b32; p[0x07] = b07; p[0x08] = b08;

    uint8_t e32 = (uint8_t)((b32 + b07 + 0xFF) & 0xFF);
    e32 = (uint8_t)((e32 + b08 + 0xF9) & 0xFF);

    obd_dtc_row_update_0x64258();

    if (p[0x32] != e32 || p[0x07] != 1 || p[0x08] != 7) {
        printf("FAIL: row=%u (b32,b07,b08)=(%02X,%02X,%02X) -> (%02X,%02X,%02X) "
               "expected (%02X,01,07)\n",
               row, b32, b07, b08, p[0x32], p[0x07], p[0x08], e32);
        return 1;
    }
    return 0;
}

int main(void)
{
    unsigned failures = 0, tests = 0;

    map_page(0xFFFF8000);

    printf("=== obd_dtc_row_update_0x64258 ===\n");

    static const uint8_t vals[] = {0x00, 0x01, 0x06, 0x07, 0x08, 0x7F, 0x80, 0xFF};
    for (uint16_t row = 0; row < 4; row++)
        for (size_t a = 0; a < sizeof(vals); a++)
            for (size_t b = 0; b < sizeof(vals); b++)
                for (size_t c = 0; c < sizeof(vals); c++) {
                    tests++;
                    failures += check(row, vals[a], vals[b], vals[c]);
                }

    srand(42);
    for (int i = 0; i < 20000; i++) {
        uint16_t row = (uint16_t)(rand() & 0x4);   /* keep p in the mapped page */
        tests++;
        failures += check(row, (uint8_t)rand(), (uint8_t)rand(), (uint8_t)rand());
    }

    printf("Results: %u tests, %u failures\n", tests, failures);
    return failures ? 1 : 0;
}
